src/Compilers/CSharp/Portable/Binder/WithExternAndUsingAliasesBinder.cs

```diff
@@ -0,0 +1,255 @@
+    {
+        private ImportChain? _lazyImportChain;
+
+        protected WithExternAndUsingAliasesBinder(WithUsingNamespacesAndTypesBinder next)
+            : base(next)
+        {
+            Debug.Assert(!next.WithImportChainEntry);
+        }
+
+        internal abstract override ImmutableArray<AliasAndUsingDirective> UsingAliases { get; }
+
+        protected abstract ImmutableDictionary<string, AliasAndUsingDirective> GetUsingAliasesMap(ConsList<TypeSymbol>? basesBeingResolved);
+
+        internal override void LookupSymbolsInSingleBinder(
+            LookupResult result, string name, int arity, ConsList<TypeSymbol> basesBeingResolved, LookupOptions options, Binder originalBinder, bool diagnose, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            Debug.Assert(result.IsClear);
+
+            LookupSymbolInAliases(
+                this.GetUsingAliasesMap(basesBeingResolved),
+                this.ExternAliases,
+                originalBinder,
+                result,
+                name,
+                arity,
+                basesBeingResolved,
+                options,
+                diagnose,
+                ref useSiteInfo);
+        }
+
+        protected override void AddLookupSymbolsInfoInSingleBinder(LookupSymbolsInfo result, LookupOptions options, Binder originalBinder)
+        {
+            // If we are looking only for labels we do not need to search through the imports.
+            if ((options & LookupOptions.LabelsOnly) == 0)
+            {
+                AddLookupSymbolsInfoInAliases(
+                    this.GetUsingAliasesMap(basesBeingResolved: null),
+                    this.ExternAliases,
+                    result, options, originalBinder);
+            }
+        }
+
+        internal override ImportChain ImportChain
+        {
+            get
+            {
+                if (_lazyImportChain == null)
+                {
+                    Debug.Assert(this.Next is WithUsingNamespacesAndTypesBinder);
+                    Interlocked.CompareExchange(ref _lazyImportChain, BuildImportChain(), null);
+                }
+
+                Debug.Assert(_lazyImportChain != null);
+
+                return _lazyImportChain;
+            }
+        }
+
+        protected abstract ImportChain BuildImportChain();
+
+        internal bool IsUsingAlias(string name, bool callerIsSemanticModel, ConsList<TypeSymbol>? basesBeingResolved)
+        {
+            return IsUsingAlias(this.GetUsingAliasesMap(basesBeingResolved), name, callerIsSemanticModel);
+        }
+
+        /// <summary>
+        /// This overload is added primarily to shadow the one from the base.
+        /// </summary>
+        [Obsolete]
+        internal static new WithExternAndUsingAliasesBinder Create(SourceNamespaceSymbol declaringSymbol, CSharpSyntaxNode declarationSyntax, Binder next)
+        {
+            return new FromSyntax(declaringSymbol, declarationSyntax, (WithUsingNamespacesAndTypesBinder)next);
```

Comment:
Consider throwing instead

Reply:
Is the implementation necessary or used?

---
In reply to: [588799194](https://github.com/dotnet/roslyn/pull/51676#discussion_r588799194) [](ancestors = 588799194)

Reply:
>Is the implementation necessary or used?

At the moment the implementation is not used, but it is correct implementation and those getting an obsolete error for this method will be able to get pretty good idea what they are required to do. 

---
In reply to: [589587752](https://github.com/dotnet/roslyn/pull/51676#discussion_r589587752) [](ancestors = 589587752,588799194)